File: server/csharp/MetaObjects.Codegen/Generators/RenderHelperGenerator.cs
// render-helper-generator — for each `template.output` declaration, emits a
// `<TemplateName>.render.cs` file declaring a static `<TemplateName>RenderHelper`
// class with a typed `Render(payload, provider)` entry point that wraps the
// EXISTING MetaObjects.Render `Renderer.Render(...)` engine — AND enforces the
// mustache↔payload-VO drift check (the EXISTING `Verify.Check`) at BUILD time.
//
// Two shapes keyed off @kind:
//   • document (default) → renders @textRef in @format → one string.
//   • email              → renders @subjectRef (text) + @htmlBodyRef (html)
//                           (+ optional @textBodyRef, text) → an EmailDocument.
//
// THE HEADLINE is the BUILD-TIME drift gate. BEFORE emitting, every referenced
// mustache (document: @textRef; email: subject + html (+ optional text) part-refs)
// is resolved through a FilesystemProvider rooted at the templateRoot ctor arg and
// run through Verify.Check. If a referenced text is unresolvable OR carries any
// NON-warning error (i.e. anything other than ERR_REQUIRED_SLOT_UNUSED — a
// `{{field}}` not on the payload VO produces ERR_VAR_NOT_ON_PAYLOAD), this THROWS
// (fails codegen), naming the template, the ref, the error code, and the offending
// field. This is what makes the build fail when a mustache references a field the
// payload VO doesn't declare.
//
// Reuse, not reimplementation: Renderer.Render (the emitted runtime call), Verify
// (the build-time gate), FilesystemProvider (build-time ref resolution), and
// EmailDocument (email return type). The payload field tree is walked from the VO
// the same way the other generators walk it — but a nested field.object's
// @objectRef is resolved by BARE short-name (cross-port render-helper consensus:
// TS findObject / Java resolveNestedObjectRef), only recursing into object.value
// (SUBTYPE_VALUE) targets, cycle-guarded.
//
// The emitted <fieldTree> literal is baked into the RenderRequest.Verify argument
// so Renderer's runtime drift check matches the build-time gate that ran here.
//
// Skips (same contract as OutputParserGenerator / OutputPromptGenerator): missing
// @payloadRef, or a @payloadRef that doesn't resolve to an object.value.
//
// Mirrors the TS port (render-helper-file.ts + templates/render-helper.ts) and the
// Java port (SpringRenderHelperGenerator). The drift message style matches exactly:
//   render-helper drift: template "<Name>" ref "<ref>" — <CODE>: {{<field>}} not on payload VO

using System.Text;
using MetaObjects.Meta;
using MetaObjects.Render;
using static MetaObjects.Shared.BaseTypes;
using static MetaObjects.Core.Field.FieldConstants;
using static MetaObjects.Core.Object.ObjectConstants;
using static MetaObjects.Template.TemplateConstants;

namespace MetaObjects.Codegen.Generators;

/// <summary>
/// Emits one <c>&lt;Template&gt;RenderHelper</c> class per <c>template.output</c>,
/// wrapping <see cref="Renderer"/> with a typed <c>Render(payload, provider)</c>
/// entry point and enforcing the mustache↔VO drift check (<see cref="Verify"/>) at
/// BUILD time. Construct with the on-disk template root the drift gate resolves each
/// referenced mustache against.
/// </summary>
public class RenderHelperGenerator : IGenerator
{
    /// <summary>
    /// The on-disk template resolver the build-time drift gate runs each referenced
    /// mustache through. <c>protected</c> so a subclass overriding <see cref="EmitHelper"/>
    /// (or the gate) can reuse the same provider.
    /// </summary>
    protected readonly FilesystemProvider _provider;

    /// <param name="templateRoot">
    /// On-disk template dir the build-time drift gate resolves each referenced
    /// mustache against (the same texts <see cref="Renderer"/> will resolve at
    /// runtime). Required — without it the drift gate cannot run.
    /// </param>
    public RenderHelperGenerator(string templateRoot)
    {
        if (string.IsNullOrEmpty(templateRoot))
            throw new ArgumentException(
                "RenderHelperGenerator requires a templateRoot (the on-disk template dir) for the build-time drift gate",
                nameof(templateRoot));
        _provider = new FilesystemProvider(templateRoot);
    }

    public virtual string Name => "render-helper-generator";

    public virtual IEnumerable<EmittedFile> Generate(GenContext ctx)
    {
        // ADR-0039: Children() — resolving root scan (behavior-identical; root has no super).
        var outputs = ctx.Root.Children()
            .Where(c => c.Type == TYPE_TEMPLATE && c.SubType == TEMPLATE_SUBTYPE_OUTPUT)
            .OrderBy(t => t.Name, StringComparer.Ordinal)
            .ToList();

        var files = new List<EmittedFile>();
        foreach (var tmpl in outputs)
        {
            // ADR-0039: resolving — @payloadRef may be inherited via an abstract template base.
            if (tmpl.Attr(TEMPLATE_ATTR_PAYLOAD_REF) is not string payloadRef)
            {
                ctx.Warn($"{Name}: template.output \"{tmpl.Name}\" missing @payloadRef — skipped.");
                continue;
            }
            // @payloadRef must resolve to an object.value (same contract as the
            // parser/prompt generators). The VO record name == payloadRef (PayloadCodegen).
            var vo = ResolveValueObject(ctx.Root, payloadRef);
            if (vo is null) continue;

            // EmitHelper runs the build-time drift gate first and THROWS (fails
            // codegen) on a mustache↔VO drift — intentionally NOT caught.
            files.Add(EmitHelper(tmpl, vo, payloadRef, ctx));
        }
        return files;
    }

    /// <summary>The root-level <c>object.value</c> a <c>@payloadRef</c> names, or null.</summary>
    internal static MetaData? ResolveValueObject(MetaRoot root, string payloadRef) =>
        // ADR-0039: Children() — resolving root scan (behavior-identical; root has no super).
        root.Children().FirstOrDefault(c =>
            c.Type == TYPE_OBJECT && c.SubType == OBJECT_SUBTYPE_VALUE && c.Name == payloadRef);

    /// <summary>
    /// True iff this generator emits a render helper for <paramref name="tmpl"/>: a
    /// <c>template.output</c> whose <c>@payloadRef</c> resolves to a root-level
    /// <c>object.value</c>. Single source of truth shared by the generator loop AND the
    /// api-docs builder (so docs never claim a suppressed symbol).
    /// </summary>
    public static bool AppliesTo(MetaData tmpl, MetaRoot root)
    {
        if (tmpl.Type != TYPE_TEMPLATE || tmpl.SubType != TEMPLATE_SUBTYPE_OUTPUT) return false;
        // ADR-0039: resolving — @payloadRef may be inherited via an abstract template base.
        if (tmpl.Attr(TEMPLATE_ATTR_PAYLOAD_REF) is not string payloadRef) return false;
        return ResolveValueObject(root, payloadRef) is not null;
    }

    protected virtual EmittedFile EmitHelper(MetaData tmpl, MetaData vo, string payloadRef, GenContext ctx)
    {
        var templateName = tmpl.Name;
        var helperClass = CSharpNaming.RenderHelperName(templateName);
        var payloadType = payloadRef; // PayloadCodegen names the record after the VO name.

        // Payload field tree — reused by the build-time gate AND baked into the
        // emitted RenderRequest.Verify so the runtime check matches the gate.
        var fields = DerivePayloadFieldTree(ctx.Root, vo, new HashSet<string>(StringComparer.Ordinal));
        var ft = FieldTreeLiteral(fields);

        // ADR-0039: resolving — @kind may be inherited via an abstract template base.
        var kind = (tmpl.Attr(TEMPLATE_ATTR_KIND) as string ?? TEMPLATE_KIND_DEFAULT).ToLowerInvariant();

        var sb = new StringBuilder();
        sb.AppendLine("// <auto-generated/>");
        sb.AppendLine("// Generated by MetaObjects render-helper-generator. Do not edit by hand.");
        sb.AppendLine("// Typed render helper wrapping the render engine; the payload field tree is");
        sb.AppendLine("// baked into the RenderRequest so the runtime drift check matches the build-time gate.");
        sb.AppendLine("#nullable enable");
        sb.AppendLine();
        sb.AppendLine($"namespace {ctx.Config.Namespace};");
        sb.AppendLine();
        sb.AppendLine($"/// <summary>Typed render helper for the <c>{templateName}</c> template.output.</summary>");
        sb.AppendLine($"public static class {helperClass}");
        sb.AppendLine("{");

        if (kind == TEMPLATE_KIND_EMAIL)
        {
            // ADR-0039: resolving — email part-refs may be inherited via an abstract template base.
            var subjectRef = tmpl.Attr(TEMPLATE_ATTR_SUBJECT_REF) as string;
            var htmlBodyRef = tmpl.Attr(TEMPLATE_ATTR_HTML_BODY_REF) as string;
            var textBodyRef = tmpl.Attr(TEMPLATE_ATTR_TEXT_BODY_REF) as string;
            if (subjectRef is null)
                throw new InvalidOperationException($"template.output \"{templateName}\" (email) missing @subjectRef");
            if (htmlBodyRef is null)
                throw new InvalidOperationException($"template.output \"{templateName}\" (email) missing @htmlBodyRef");

            // BUILD-TIME drift gate — every email part-ref is resolved + verified.
            GateRef(templateName, subjectRef, fields);
            GateRef(templateName, htmlBodyRef, fields);
            if (textBodyRef is not null) GateRef(templateName, textBodyRef, fields);

            sb.AppendLine($"    /// <summary>Render the {templateName} email (subject + html body{(textBodyRef is not null ? " + text body" : "")}) from a typed {payloadType} payload.</summary>");
            sb.AppendLine($"    public static global::MetaObjects.Render.EmailDocument Render({payloadType} payload, global::MetaObjects.Render.IProvider provider) =>");
            sb.AppendLine("        new global::MetaObjects.Render.EmailDocument(");
            sb.AppendLine($"            global::MetaObjects.Render.Renderer.Render(new global::MetaObjects.Render.RenderRequest {{ Ref = {Quote(subjectRef)}, Payload = payload, Provider = provider, Format = \"text\", Verify = {ft} }}),");
            sb.Append($"            global::MetaObjects.Render.Renderer.Render(new global::MetaObjects.Render.RenderRequest {{ Ref = {Quote(htmlBodyRef)}, Payload = payload, Provider = provider, Format = \"html\", Verify = {ft} }})");
            sb.AppendLine(",");
            if (textBodyRef is not null)
                sb.AppendLine($"            global::MetaObjects.Render.Renderer.Render(new global::MetaObjects.Render.RenderRequest {{ Ref = {Quote(textBodyRef)}, Payload = payload, Provider = provider, Format = \"text\", Verify = {ft} }}));");
            else
                sb.AppendLine("            null);");
        }
        else
        {
            // ADR-0039: resolving — @textRef/@format may be inherited via an abstract template base.
            var textRef = tmpl.Attr(TEMPLATE_ATTR_TEXT_REF) as string;
            if (textRef is null)
                throw new InvalidOperationException($"template.output \"{templateName}\" (document) missing @textRef");
            var format = tmpl.Attr(TEMPLATE_ATTR_FORMAT) as string ?? TEMPLATE_FORMAT_DEFAULT;
            var maxChars = MaxCharsOf(tmpl);

            // BUILD-TIME drift gate.
            GateRef(templateName, textRef, fields);

            var maxCharsArg = maxChars is not null ? $", MaxChars = {maxChars}" : "";
            sb.AppendLine($"    /// <summary>Render the {templateName} document from a typed {payloadType} payload.</summary>");
            sb.AppendLine($"    public static string Render({payloadType} payload, global::MetaObjects.Render.IProvider provider) =>");
            sb.AppendLine($"        global::MetaObjects.Render.Renderer.Render(new global::MetaObjects.Render.RenderRequest {{ Ref = {Quote(textRef)}, Payload = payload, Provider = provider, Format = {Quote(format)}, Verify = {ft}{maxCharsArg} }});");
        }

        sb.AppendLine("}");
        return new EmittedFile($"{templateName}.render.cs", sb.ToString());
    }

    // -------------------------------------------------------------------------
    // Build-time drift gate
    // -------------------------------------------------------------------------

    /// <summary>
    /// Run the build-time drift gate for one referenced mustache. THROWS (fails
    /// codegen) when the ref is unresolvable OR <see cref="Verify.Check"/> reports a
    /// non-warning error. Warnings (<see cref="Verify.ERR_REQUIRED_SLOT_UNUSED"/>)
    /// are tolerated. Matches the TS/Java message style exactly.
    /// </summary>
    private void GateRef(string templateName, string reference, IReadOnlyList<PayloadField> fields)
    {
        var text = _provider.Resolve(reference);
        if (text is null)
            throw new InvalidOperationException(
                $"render-helper drift: template \"{templateName}\" ref \"{reference}\" — unresolved (provider returned no text)");

        var opts = new VerifyOptions { Provider = _provider };
        foreach (var e in Verify.Check(text, fields, opts))
        {
            if (e.Code == Verify.ERR_REQUIRED_SLOT_UNUSED) continue; // warning
            throw new InvalidOperationException(
                $"render-helper drift: template \"{templateName}\" ref \"{reference}\" — {e.Code}: {{{{{e.Path}}}}} not on payload VO");
        }
    }

    // -------------------------------------------------------------------------
    // Payload field-tree walk — nested @objectRef resolved by BARE short-name
    // (cross-port render-helper consensus: TS findObject / Java
    // resolveNestedObjectRef). Object-ref fields recurse into their target
    // object.value (SUBTYPE_VALUE only); a `seen` set guards reference cycles.
    // -------------------------------------------------------------------------

    private static IReadOnlyList<PayloadField> DerivePayloadFieldTree(
        MetaRoot root, MetaData vo, HashSet<string> seen)
    {
        if (vo is null || !seen.Add(vo.Name)) return [];
        var fields = new List<PayloadField>();
        foreach (var f in vo.Children().Where(c => c.Type == TYPE_FIELD))
        {
            if (f.SubType == FIELD_SUBTYPE_OBJECT &&
                // ADR-0039: resolving — @objectRef may be inherited via extends (TS reads f.attr).
                f.Attr(FIELD_ATTR_OBJECT_REF) is string refName)
            {
                var target = ResolveNestedObjectRef(root, refName);
                if (target is not null && target.SubType == OBJECT_SUBTYPE_VALUE)
                {
                    var children = DerivePayloadFieldTree(root, target, new HashSet<string>(seen, StringComparer.Ordinal));
                    fields.Add(new PayloadField(f.Name, children));
                    continue;
                }
            }
            fields.Add(new PayloadField(f.Name));
        }
        return fields;
    }

    /// <summary>
    /// Resolve a <c>field.object</c>'s <c>@objectRef</c> to its target
    /// <c>object.value</c> by BARE short-name — mirroring the TS render-helper's
    /// <c>findObject(root, ref)</c> (<c>c.type === OBJECT &amp;&amp; c.name === ref</c>)
    /// and the Java <c>resolveNestedObjectRef</c>. If the ref carries a package, only
    /// the segment after the last <c>::</c> is compared, against each
    /// <c>object.value</c>'s own short name. Returns null when no match is found.
    /// </summary>
    private static MetaData? ResolveNestedObjectRef(MetaRoot root, string reference)
    {
        if (string.IsNullOrEmpty(reference)) return null;
        var refShort = CSharpNaming.StripPkg(reference);
        // ADR-0039: Children() — resolving root scan (behavior-identical; root has no super).
        return root.Children().FirstOrDefault(c =>
            c.Type == TYPE_OBJECT && c.SubType == OBJECT_SUBTYPE_VALUE &&
            CSharpNaming.StripPkg(c.Name) == refShort);
    }

    /// <summary>
    /// Emit a <c>IReadOnlyList&lt;PayloadField&gt;</c> as a deterministic C# array
    /// literal: <c>new PayloadField[] { new PayloadField("name"),
    /// new PayloadField("nested", new PayloadField[] { ... }) }</c>. An empty tree
    /// emits <c>new global::MetaObjects.Render.PayloadField[] { }</c>.
    /// </summary>
    private static string FieldTreeLiteral(IReadOnlyList<PayloadField> fields)
    {
        var sb = new StringBuilder("new global::MetaObjects.Render.PayloadField[] { ");
        for (int i = 0; i < fields.Count; i++)
        {
            if (i > 0) sb.Append(", ");
            var f = fields[i];
            if (f.Fields is not null)
                sb.Append($"new global::MetaObjects.Render.PayloadField({Quote(f.Name)}, {FieldTreeLiteral(f.Fields)})");
            else
                sb.Append($"new global::MetaObjects.Render.PayloadField({Quote(f.Name)})");
        }
        sb.Append(" }");
        return sb.ToString();
    }

    // -------------------------------------------------------------------------
    // Local helpers
    // -------------------------------------------------------------------------

    /// <summary>Resolve @maxChars as an int, or null when absent/blank/non-numeric.
    /// ADR-0039: resolving — @maxChars may be inherited via an abstract template base.</summary>
    private static int? MaxCharsOf(MetaData tmpl)
    {
        var v = tmpl.Attr(TEMPLATE_ATTR_MAX_CHARS);
        return v switch
        {
            int i => i,
            long l => (int)l,
            string s when int.TryParse(s, out var n) => n,
            _ => null,
        };
    }

    private static string Quote(string s) =>
        "\"" + s.Replace("\\", "\\\\").Replace("\"", "\\\"") + "\"";
}

File: server/python/src/metaobjects/codegen/generators/render_helper_generator.py
"""Render-helper codegen — one ``<template_name_snake>_render_helper.py`` per
``template.output`` declaration (render-helper phase 2).

For each ``template.output`` this emits a typed ``render_<name>(payload, provider)``
function that WRAPS the existing :func:`metaobjects.render.renderer.render` engine,
and enforces the mustache↔payload-VO drift check (the existing
:func:`metaobjects.render.verify.verify`) at BUILD time.

Two shapes keyed off ``@kind``:

* ``document`` (default) → renders ``@textRef`` in ``@format`` → one ``str``.
* ``email`` → renders ``@subjectRef`` (text) + ``@htmlBodyRef`` (html)
  (+ optional ``@textBodyRef``, text) → an
  :class:`metaobjects.render.email_document.EmailDocument`.

THE HEADLINE is the BUILD-TIME drift gate. BEFORE emitting, every referenced
mustache (document: ``@textRef``; email: subject + html (+ optional text)) is
resolved through a :class:`~metaobjects.render.filesystem_provider.FilesystemProvider`
rooted at the ``template_root`` ctor arg and run through ``verify``. If a referenced
text is unresolvable OR carries any NON-warning error (anything other than
``ERR_REQUIRED_SLOT_UNUSED`` — a ``{{field}}`` not on the payload VO produces
``ERR_VAR_NOT_ON_PAYLOAD``), the generator RAISES (fails codegen), naming the
template, the ref, the error code, and the offending field. This is what makes the
build fail when a mustache references a field the payload VO doesn't declare.

Reuse, not reimplementation: ``render`` (the emitted runtime call), ``verify`` (the
build-time gate), ``FilesystemProvider`` (build-time ref resolution), and
``EmailDocument`` (email return type). The payload field tree is walked from the VO
the same way the other generators walk it — a nested ``field.object``'s ``@objectRef``
is resolved FQN-exact when fully-qualified (ADR-0041) else by short name (cross-port
render-helper consensus: TS ``refMatchesObject`` / Java+Kotlin ``resolveNestedObjectRef``
/ C# ``ResolveNestedObjectRef``), only recursing into ``object.value`` targets, cycle-guarded.

Python divergence vs TS / Java / C#: Python's ``RenderRequest`` has NO ``verify``
field — the Python render engine does not run a runtime drift pass — so the emitted
helper does NOT pass a runtime ``verify`` field-tree. The BUILD-TIME gate that runs
here is the only, and a complete, drift guarantee: codegen cannot succeed unless
every referenced mustache verifies clean against the payload VO. The field-tree walk
is still performed (it is what ``verify`` checks against); it is simply not baked into
the emitted call. The emitted helper shape is otherwise identical to the other ports.

Drift message style matches the other ports exactly:
``render-helper drift: template "<N>" ref "<r>" — <CODE>: {{<f>}} not on payload VO``
"""
from __future__ import annotations

from collections.abc import Callable

from metaobjects.apidocs.naming import snake_case as _snake_case
from metaobjects.codegen.constants import generated_header
from metaobjects.codegen.format import ruff_format
from metaobjects.codegen.generator import EmittedFile, GenContext, Generator
from metaobjects.meta.core.field import field_constants as fc
from metaobjects.meta.core.field.meta_field import MetaField
from metaobjects.meta.core.object.meta_object import MetaObject
from metaobjects.meta.core.object.object_constants import OBJECT_SUBTYPE_VALUE
from metaobjects.meta.meta_data import MetaData
from metaobjects.meta.template import template_constants as tc
from metaobjects.render.filesystem_provider import FilesystemProvider
from metaobjects.render.verify import (
    ERR_REQUIRED_SLOT_UNUSED,
    PayloadField,
    verify,
)
from metaobjects.shared.base_types import TYPE_FIELD, TYPE_OBJECT, TYPE_TEMPLATE
from metaobjects.shared.separators import PACKAGE_SEP

_GENERATOR_NAME = "render-helper-generator"


def _py_str(s: str) -> str:
    """A double-quoted Python string literal with the minimal escaping the refs/formats
    need (refs are ``group/source`` slugs; formats are closed-enum words)."""
    return '"' + s.replace("\\", "\\\\").replace('"', '\\"') + '"'


# ---------------------------------------------------------------------------
# Payload field-tree walk — nested @objectRef resolved FQN-exact (ADR-0041).
# Object-ref fields recurse into their target object.value (SUBTYPE_VALUE only);
# a `seen` set guards reference cycles.
# ---------------------------------------------------------------------------


def _resolve_nested_object_ref(root: MetaData, reference: str) -> MetaObject | None:
    """Resolve a ``field.object``'s ``@objectRef`` to its target ``object.value``.

    ADR-0041: a FULLY-QUALIFIED ref (contains ``::``) resolves EXACTLY on the
    package-qualified name (``resolution_key()``/``fqn()``) — never a bare-tail
    fallback that would bind a same-named ``object.value`` in the WRONG package on a
    cross-package short-name collision. A bare ref still matches by short name
    (first-wins). Mirrors the Java/Kotlin ``resolveNestedObjectRef`` + TS
    ``refMatchesObject``."""
    if not reference:
        return None
    fqn = PACKAGE_SEP in reference
    ref_short = reference.rsplit(PACKAGE_SEP, 1)[-1]
    # ADR-0039 sanctioned own: top-level scan on the loader ROOT (never extended, own == effective)
    for child in root.own_children():
        if child.type != TYPE_OBJECT or not isinstance(child, MetaObject):
            continue
        if child.sub_type != OBJECT_SUBTYPE_VALUE:
            continue
        if fqn:
            if child.resolution_key() == reference or child.fqn() == reference:
                return child
        elif child.name.rsplit(PACKAGE_SEP, 1)[-1] == ref_short:
            return child
    return None


def _derive_payload_field_tree(
    root: MetaData, vo: MetaObject, seen: frozenset[str]
) -> list[PayloadField]:
    """Walk *vo*'s fields into a :class:`PayloadField` tree. A ``field.object`` with a
    resolvable ``@objectRef`` to an ``object.value`` becomes a context-pushing node whose
    children are the target VO's tree (recursed, cycle-guarded). Every other field is a
    leaf."""
    if vo is None or vo.resolution_key() in seen:
        return []
    next_seen = seen | {vo.resolution_key()}
    fields: list[PayloadField] = []
    for f in vo.children():
        if f.type != TYPE_FIELD or not isinstance(f, MetaField):
            continue
        if f.sub_type == fc.FIELD_SUBTYPE_OBJECT:
            ref = f.attrs().get(fc.FIELD_ATTR_OBJECT_REF)
            if isinstance(ref, str) and ref:
                target = _resolve_nested_object_ref(root, ref)
                if target is not None and target.sub_type == OBJECT_SUBTYPE_VALUE:
                    children = _derive_payload_field_tree(root, target, next_seen)
                    fields.append(PayloadField(f.name, children))
                    continue
        fields.append(PayloadField(f.name))
    return fields


def _field_tree_literal(fields: list[PayloadField]) -> str:
    """Emit a ``list[PayloadField]`` as a deterministic Python literal:
    ``[PayloadField("name"), PayloadField("nested", [PayloadField("x")])]``."""
    parts: list[str] = []
    for f in fields:
        if f.fields is not None:
            parts.append(f"PayloadField({_py_str(f.name)}, {_field_tree_literal(f.fields)})")
        else:
            parts.append(f"PayloadField({_py_str(f.name)})")
    return "[" + ", ".join(parts) + "]"


# ---------------------------------------------------------------------------
# Resolution + emission.
# ---------------------------------------------------------------------------


def _resolve_payload_vo(root: MetaData, payload_ref: str) -> MetaObject | None:
    """``@payloadRef`` must resolve to an ``object.value`` (same contract as the
    parser/prompt generators). Matches a value-object child by short name OR by the
    last segment of a fully-qualified ``a::b::Name`` ref (FR-026 expands refs to FQN,
    while the child node still carries the short ``name``)."""
    ref_short = payload_ref.rsplit("::", 1)[-1]
    # ADR-0039 sanctioned own: top-level scan on the loader ROOT (never extended, own == effective)
    for child in root.own_children():
        if child.type != TYPE_OBJECT or not isinstance(child, MetaObject):
            continue
        if child.sub_type == OBJECT_SUBTYPE_VALUE and child.name in (payload_ref, ref_short):
            return child
    return None


def _max_chars_of(tmpl: MetaData) -> int | None:
    """Resolve ``@maxChars`` (own attr) as an int, or ``None`` when absent/non-numeric."""
    v = tmpl.get_meta_attr(tc.TEMPLATE_ATTR_MAX_CHARS)  # ADR-0039: template attr resolves via extends (not origin; templates CAN extend)
    if isinstance(v, bool):
        return None
    if isinstance(v, int):
        return v
    if isinstance(v, str):
        try:
            return int(v)
        except ValueError:
            return None
    return None


class RenderHelperGenerator:
    """Generator wrapping the per-``template.output`` render-helper emit. Construct with
    the on-disk template root the build-time drift gate resolves each referenced
    mustache against — required, without it the gate cannot run.

    EXTENSION SEAM (open-for-extension). Adopters subclass this and override one of
    the protected emit hooks to customize the emitted helper without forking:

    * ``_emit_document(header, template_name, snake, payload_ref, tmpl, fields)`` —
      the ``document``-kind helper (single ``str`` return).
    * ``_emit_email(header, template_name, snake, payload_ref, tmpl, fields)`` — the
      ``email``-kind helper (``EmailDocument`` return).
    * ``_emit_helper(tmpl, vo, payload_ref, root)`` — the per-template dispatch
      (runs the build-time drift gate, then routes to document/email).

    The build-time drift gate (``_gate_ref``) is also overridable, but tightening it
    is the safer direction than loosening it. The factory ``render_helper_generator()``
    returns a default instance, so the default suite stays byte-identical."""

    name = _GENERATOR_NAME

    def __init__(
        self,
        template_root: str,
        *,
        filter: Callable[[MetaObject], bool] | None = None,
    ) -> None:
        if not template_root:
            raise ValueError(
                "RenderHelperGenerator requires a template_root (the on-disk template "
                "dir) for the build-time drift gate"
            )
        # The ``filter`` arg matches the cross-generator contract even though this
        # generator iterates templates (not entities).
        self.filter = filter
        self._provider = FilesystemProvider(template_root)

    def generate(self, ctx: GenContext) -> list[EmittedFile]:
        root = ctx.loaded_root
        if root is None:
            return []
        outputs = sorted(
            (
                # ADR-0039 sanctioned own: top-level scan on the loader ROOT (never extended, own == effective)
                c
                for c in root.own_children()
                if c.type == TYPE_TEMPLATE
                and c.sub_type == tc.TEMPLATE_SUBTYPE_OUTPUT
            ),
            key=lambda c: c.name,
        )
        files: list[EmittedFile] = []
        for tmpl in outputs:
            payload_ref = tmpl.get_meta_attr(tc.TEMPLATE_ATTR_PAYLOAD_REF)  # ADR-0039: template attr resolves via extends (not origin; templates CAN extend)
            if not isinstance(payload_ref, str) or not payload_ref:
                ctx.warn(
                    f"{_GENERATOR_NAME}: template.output '{tmpl.name}' missing "
                    "@payloadRef — skipped."
                )
                continue
            vo = _resolve_payload_vo(root, payload_ref)
            if vo is None:
                ctx.warn(
                    f"{_GENERATOR_NAME}: template.output '{tmpl.name}' @payloadRef "
                    f"'{payload_ref}' does not resolve to an object.value — skipped."
                )
                continue
            # _emit_helper runs the build-time drift gate first and RAISES (fails
            # codegen) on a mustache↔VO drift — intentionally NOT caught.
            content = self._emit_helper(tmpl, vo, payload_ref, root)
            files.append(
                EmittedFile(
                    path=f"{_snake_case(tmpl.name)}_render_helper.py",
                    content=ruff_format(content),
                )
            )
        return files

    # -- build-time drift gate -------------------------------------------------

    def _gate_ref(
        self, template_name: str, reference: str, fields: list[PayloadField]
    ) -> None:
        """Run the build-time drift gate for one referenced mustache. RAISES (fails
        codegen) when the ref is unresolvable OR ``verify`` reports a non-warning error.
        Warnings (``ERR_REQUIRED_SLOT_UNUSED``) are tolerated. Matches the cross-port
        message style exactly."""
        text = self._provider.resolve(reference)
        if text is None:
            raise ValueError(
                f'render-helper drift: template "{template_name}" ref "{reference}" '
                "— unresolved (provider returned no text)"
            )
        for e in verify(text, fields, provider=self._provider):
            if e.code == ERR_REQUIRED_SLOT_UNUSED:
                continue  # warning
            raise ValueError(
                f'render-helper drift: template "{template_name}" ref "{reference}" '
                f"— {e.code}: {{{{{e.path}}}}} not on payload VO"
            )

    # -- emission --------------------------------------------------------------

    def _emit_helper(
        self,
        tmpl: MetaData,
        vo: MetaObject,
        payload_ref: str,
        root: MetaData,
    ) -> str:
        template_name = tmpl.name
        snake = _snake_case(template_name)
        fields = _derive_payload_field_tree(root, vo, frozenset())
        kind = tmpl.get_meta_attr(tc.TEMPLATE_ATTR_KIND)  # ADR-0039: template attr resolves via extends (not origin; templates CAN extend)
        kind = (kind if isinstance(kind, str) else tc.TEMPLATE_KIND_DEFAULT).lower()

        fqn = (
            f"{vo.package}{PACKAGE_SEP}{template_name}"
            if getattr(vo, "package", None)
            else template_name
        )
        header = generated_header(template_name, fqn)

        if kind == tc.TEMPLATE_KIND_EMAIL:
            return self._emit_email(header, template_name, snake, payload_ref, tmpl, fields)
        return self._emit_document(header, template_name, snake, payload_ref, tmpl, fields)

    def _emit_document(
        self,
        header: str,
        template_name: str,
        snake: str,
        payload_ref: str,
        tmpl: MetaData,
        fields: list[PayloadField],
    ) -> str:
        text_ref = tmpl.get_meta_attr(tc.TEMPLATE_ATTR_TEXT_REF)  # ADR-0039: template attr resolves via extends (not origin; templates CAN extend)
        if not isinstance(text_ref, str) or not text_ref:
            raise ValueError(
                f'template.output "{template_name}" (document) missing @textRef'
            )
        fmt = tmpl.get_meta_attr(tc.TEMPLATE_ATTR_FORMAT)  # ADR-0039: template attr resolves via extends (not origin; templates CAN extend)
        fmt = fmt if isinstance(fmt, str) and fmt else tc.TEMPLATE_FORMAT_DEFAULT
        max_chars = _max_chars_of(tmpl)

        # BUILD-TIME drift gate (THE headline) — raises on drift.
        self._gate_ref(template_name, text_ref, fields)

        request_lines = [
            "        RenderRequest(",
            "            payload=payload,",
            "            provider=provider,",
            f"            ref={_py_str(text_ref)},",
            f"            format={_py_str(fmt)},",
        ]
        if max_chars is not None:
            request_lines.append(f"            max_chars={max_chars},")
        request_lines.append("        )")

        lines = [
            header,
            "from __future__ import annotations",
            "",
            "from metaobjects.render.renderer import render, RenderRequest",
            "",
            "",
            f"def render_{snake}(payload, provider) -> str:",
            f'    """Render the ``{template_name}`` document from a typed '
            f"``{payload_ref}`` payload.",
            "",
            "    Wraps the render engine; the mustache↔payload-VO drift check ran at",
            '    BUILD time (codegen fails on drift)."""',
            "    return render(",
            *request_lines,
            "    )",
            "",
            "",
            f'__all__ = ["render_{snake}"]',
            "",
        ]
        return "\n".join(lines)

    def _emit_email(
        self,
        header: str,
        template_name: str,
        snake: str,
        payload_ref: str,
        tmpl: MetaData,
        fields: list[PayloadField],
    ) -> str:
        subject_ref = tmpl.get_meta_attr(tc.TEMPLATE_ATTR_SUBJECT_REF)  # ADR-0039: template attr resolves via extends (not origin; templates CAN extend)
        html_body_ref = tmpl.get_meta_attr(tc.TEMPLATE_ATTR_HTML_BODY_REF)  # ADR-0039: template attr resolves via extends (not origin; templates CAN extend)
        text_body_ref = tmpl.get_meta_attr(tc.TEMPLATE_ATTR_TEXT_BODY_REF)  # ADR-0039: template attr resolves via extends (not origin; templates CAN extend)
        if not isinstance(subject_ref, str) or not subject_ref:
            raise ValueError(
                f'template.output "{template_name}" (email) missing @subjectRef'
            )
        if not isinstance(html_body_ref, str) or not html_body_ref:
            raise ValueError(
                f'template.output "{template_name}" (email) missing @htmlBodyRef'
            )
        has_text = isinstance(text_body_ref, str) and bool(text_body_ref)

        # BUILD-TIME drift gate — every email part-ref is resolved + verified.
        self._gate_ref(template_name, subject_ref, fields)
        self._gate_ref(template_name, html_body_ref, fields)
        if has_text:
            self._gate_ref(template_name, text_body_ref, fields)

        text_body_expr = (
            "render(\n"
            "            RenderRequest(\n"
            "                payload=payload,\n"
            "                provider=provider,\n"
            f"                ref={_py_str(text_body_ref)},\n"
            '                format="text",\n'
            "            )\n"
            "        )"
            if has_text
            else "None"
        )

        lines = [
            header,
            "from __future__ import annotations",
            "",
            "from metaobjects.render.email_document import EmailDocument",
            "from metaobjects.render.renderer import render, RenderRequest",
            "",
            "",
            f"def render_{snake}(payload, provider) -> EmailDocument:",
            f'    """Render the ``{template_name}`` email (subject + html body'
            f'{" + text body" if has_text else ""}) from a typed ``{payload_ref}`` payload.',
            "",
            "    Wraps the render engine; the mustache↔payload-VO drift check ran at",
            '    BUILD time (codegen fails on drift)."""',
            "    return EmailDocument(",
            "        subject=render(",
            "            RenderRequest(",
            "                payload=payload,",
            "                provider=provider,",
            f"                ref={_py_str(subject_ref)},",
            '                format="text",',
            "            )",
            "        ),",
            "        html_body=render(",
            "            RenderRequest(",
            "                payload=payload,",
            "                provider=provider,",
            f"                ref={_py_str(html_body_ref)},",
            '                format="html",',
            "            )",
            "        ),",
            f"        text_body={text_body_expr},",
            "    )",
            "",
            "",
            f'__all__ = ["render_{snake}"]',
            "",
        ]
        return "\n".join(lines)


def render_helper_generator(
    *,
    template_root: str,
    filter: Callable[[MetaObject], bool] | None = None,
) -> Generator:
    """Factory mirroring the TS ``renderHelper()`` / C# ``RenderHelperGenerator`` /
    Java ``SpringRenderHelperGenerator`` constructors."""
    return RenderHelperGenerator(template_root, filter=filter)
